File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/ai/harness/maven/RequireHookShebangRule.java
package ai.harness.maven;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Path;
import java.util.List;

/**
 * Rule that requires hook files to start with a specific shebang.
 */
public class RequireHookShebangRule implements HarnessCheckRule {
    private static final String CATEGORY = "requireHookShebang";

    @Override
    public boolean applies(JsonNode manifest) {
        return HarnessCheckHelper.applies(manifest, CATEGORY);
    }

    @Override
    public List<Finding> validate(Path root, JsonNode manifest) throws MojoExecutionException {
        JsonNode catNode = manifest.get(CATEGORY);
        List<String> hooks = HarnessCheckHelper.extractPaths(catNode.get("parameters").get("hooks"));
        String expected = catNode.get("parameters").get("expectedShebang").asText();
        String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        return hooks.stream()
                .flatMap(hook -> validateHookShebang(root, hook, expected, severity).stream())
                .toList();
    }

    private List<Finding> validateHookShebang(Path root, String hook, String expected, String severity) throws MojoExecutionException {
        Path hookPath = root.resolve(hook);
        return HarnessCheckHelper.isSafeRegularFile(root, hookPath) && !HarnessCheckHelper.readFile(root, hookPath).startsWith(expected)
                ? List.of(new Finding(severity, CATEGORY, hook + " must start with " + expected))
                : List.of();
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/ai/harness/gradle/RequireHookShebangRule.kt
package ai.harness.gradle

import kotlinx.serialization.json.JsonObject
import java.nio.file.Path
import kotlin.io.path.div
import kotlin.io.path.isRegularFile
import java.io.File

/**
 * Rule that requires hooks to have correct shebang.
 */
class RequireHookShebangRule : HarnessCheckRule {
	override fun applies(manifest: JsonObject): Boolean {
		val category = "requireHookShebang"
		val catObj = manifest[category]?.jsonObject ?: return false
		val enabled = catObj["enabled"]?.jsonPrimitive?.contentOrNull?.toBoolean() ?: true
		return enabled
	}

	override fun validate(manifest: JsonObject, root: Path, psiResults: HarnessPsiResults?): Collection<Finding> {
		val category = "requireHookShebang"
		val severity = HarnessCheck.Companion.severityOf(manifest, category)
		val catObj = manifest[category]?.jsonObject ?: return emptyList()
		val parametersObj = catObj["parameters"]?.jsonObject ?: return emptyList()
		val messagesObj = catObj["messages"]?.jsonObject ?: return emptyList()
		val hooks = HarnessCheck.Companion.stringArrayFrom(parametersObj, "hooks")
		val expectedShebang = HarnessCheck.Companion.stringFrom(parametersObj, "expectedShebang")

		return buildSet<Finding> {
			hooks.forEach { hookPath ->
				val hook = root / hookPath
				if (hook.isRegularFile()) {
					val first = hook.toFile().readLines().firstOrNull() ?: ""
					if (first != expectedShebang) {
						val msg = HarnessCheck.Companion.stringFrom(messagesObj, "default").takeIf { it.isNotEmpty() } ?: "$hookPath must start with $expectedShebang"
						add(Finding(severity, category, msg))
					}
				}
			}
		}.toList()
	}
}
